<a href="https://colab.research.google.com/github/Jessica-Emereuwa/GoMyCode-DataScience/blob/main/Unsupervised_Association_Rules_Checkpoint.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
"""Instructions

toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],

        ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],

        ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],

        ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],

      ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

Run the apriori algorithm on the provided toy_dataset. Interpret the results.
Try to explore the checkpoint dataset using Pandas and Plotly.
Run the apriori algorithm on checkpoint dataset. Interpret the results and suggest a clear business plan to the supermarket owners based on your findings."""


In [7]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

# Provided toy dataset
toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],
               ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],
               ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],
               ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],
               ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [8]:
# Transform the dataset into a format suitable for the apriori algorithm
te = TransactionEncoder()
te_ary = te.fit(toy_dataset).transform(toy_dataset) #Apply one-hot-encoding on our dataset
df = pd.DataFrame(te_ary, columns=te.columns_) #Creating a nw Dataframe from our numpy array

df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Dress  Earrings    Hat  Heels  Necklace  Pants  Sandals  Scarf  Shirt  \
0  False     False   True  False     False   True    False   True  False   
1  False     False   True  False     False   True    False  False  False   
2   True     False  False   True     False   True     True   True  False   
3   True      True   True   True      True  False    False   True  False   
4  False      True  False  False     False   True    False   True   True   

   Skirt  Sneakers  Sunglasses  
0   True      True       False  
1   True      True        True  
2  False     False       False  
3  False     False       False  
4   True     False       False

In [12]:
#Try to explore the checkpoint dataset using Pandas and Plotly.
import plotly.express as px

# Calculate item frequencies
item_frequencies = df.sum()  # Calculate sum of each column (item)

# Create a DataFrame for plotting
df_plot = pd.DataFrame({'Item': item_frequencies.index, 'Frequency': item_frequencies.values})

# Bar plot of item frequencies
fig = px.bar(df_plot, x='Item', y='Frequency', title='Frequency of Items in Toy Dataset')
fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [9]:
# Apply the apriori algorithm
frequent_itemsets = apriori(df, min_support=0.4, use_colnames=True)
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                       itemsets
0       0.4                        (Dress)
1       0.4                     (Earrings)
2       0.6                          (Hat)
3       0.4                        (Heels)
4       0.8                        (Pants)
5       0.8                        (Scarf)
6       0.6                        (Skirt)
7       0.4                     (Sneakers)
8       0.4                 (Dress, Heels)
9       0.4                 (Dress, Scarf)
10      0.4              (Scarf, Earrings)
11      0.4                   (Hat, Pants)
12      0.4                   (Hat, Scarf)
13      0.4                   (Hat, Skirt)
14      0.4                (Hat, Sneakers)
15      0.4                 (Scarf, Heels)
16      0.6                 (Scarf, Pants)
17      0.6                 (Pants, Skirt)
18      0.4              (Sneakers, Pants)
19      0.4                 (Scarf, Skirt)
20      0.4              (Sneakers, Skirt)
21      0.4          (Dress, Scarf, Heels)
22      0.4            (Hat, Pants, Skirt)
23      0.4         (Hat, Sneakers, Pants)
24      0.4         (Hat, Sneakers, Skirt)
25      0.4          (Scarf, Pants, Skirt)
26      0.4       (Sneakers, Pants, Skirt)
27      0.4  (Hat, Sneakers, Pants, Skirt)

**INTERPRET THE RESULT**



*  Pants and scarves are the most frequently purchased items, appearing in 80% of transactions.

*   The combination of hat, sneakers, pants, and skirt appears in 40% of transactions.


*   Pants and scarf are frequently bought together, appearing in 60% of transactions.

In [ ]:
# Generate the association rules
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6)

In [14]:
print("Frequent Itemsets:\n", frequent_itemsets)

Frequent Itemsets:
     support                       itemsets
0       0.4                        (Dress)
1       0.4                     (Earrings)
2       0.6                          (Hat)
3       0.4                        (Heels)
4       0.8                        (Pants)
5       0.8                        (Scarf)
6       0.6                        (Skirt)
7       0.4                     (Sneakers)
8       0.4                 (Dress, Heels)
9       0.4                 (Dress, Scarf)
10      0.4              (Scarf, Earrings)
11      0.4                   (Hat, Pants)
12      0.4                   (Hat, Scarf)
13      0.4                   (Hat, Skirt)
14      0.4                (Hat, Sneakers)
15      0.4                 (Scarf, Heels)
16      0.6                 (Scarf, Pants)
17      0.6                 (Pants, Skirt)
18      0.4              (Sneakers, Pants)
19      0.4                 (Scarf, Skirt)
20      0.4              (Sneakers, Skirt)
21      0.4          (Dress, Scarf

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [18]:
print("\nAssociation Rules:\n", rules)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



antecedents               consequents  antecedent support  \
0             (Dress)                   (Heels)                 0.4   
1             (Heels)                   (Dress)                 0.4   
2             (Dress)                   (Scarf)                 0.4   
3          (Earrings)                   (Scarf)                 0.4   
4               (Hat)                   (Pants)                 0.6   
..                ...                       ...                 ...   
57  (Sneakers, Skirt)              (Hat, Pants)                 0.4   
58     (Skirt, Pants)           (Hat, Sneakers)                 0.6   
59              (Hat)  (Sneakers, Skirt, Pants)                 0.6   
60         (Sneakers)       (Hat, Skirt, Pants)                 0.4   
61            (Skirt)    (Hat, Sneakers, Pants)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  \
0                  0.4      0.4    1.000000  2.500000      0.24         inf   
1                  0.4      0.4    1.000000  2.500000      0.24         inf   
2                  0.8      0.4    1.000000  1.250000      0.08         inf   
3                  0.8      0.4    1.000000  1.250000      0.08         inf   
4                  0.8      0.4    0.666667  0.833333     -0.08         0.6   
..                 ...      ...         ...       ...       ...         ...   
57                 0.4      0.4    1.000000  2.500000      0.24         inf   
58                 0.4      0.4    0.666667  1.666667      0.16         1.8   
59                 0.4      0.4    0.666667  1.666667      0.16         1.8   
60                 0.4      0.4    1.000000  2.500000      0.24         inf   
61                 0.4      0.4    0.666667  1.666667      0.16         1.8   

    zhangs_metric  
0        1.000000  
1        1.000000  
2        0.333333  
3        0.333333  
4       -0.333333  
..            ...  
57       1.000000  
58       1.000000  
59       1.000000  
60       1.000000  
61       1.000000  

[62 rows x 10 columns]

**Interpretation of Association Rules**



*   Dress and Heels: Customers who buy dresses also buy heels with 100% confidence, indicating a strong correlation (lift of 2.5).


*   Hat, Sneakers, Pants, and Skirt: The combination of these items frequently occurs together, indicating a strong relationship (lift of 2.5).


*   Dress and Scarf: Dresses are often purchased with scarves, showing a confidence of 100% but with a lower lift (1.25).


*   Earrings and Scarf: Earrings are frequently bought with scarves, also with a confidence of 100% and a lift of 1.25.





**Business Plan Recommendations**


*   Create promotional bundles for dresses and heels. Given the strong association, customers are likely to appreciate discounts or special offers on these combinations.

*   Since dresses are frequently purchased with scarves, consider offering a discount on scarves with the purchase of a dress. This can encourage customers to buy both items.



